File: content/courses/parallel-computing-introduction/codes/bcast_ch.f90
program broadcaster
    use mpi
    implicit none

    integer           :: nprocs, rank, ierr
    integer           :: sendcount
    character(len=15) :: message

    call MPI_Init(ierr)
    call MPI_Comm_size(MPI_COMM_WORLD,nprocs,ierr)
    call MPI_Comm_rank(MPI_COMM_WORLD,rank,ierr)

    if (rank==0) then
        message="I have a secret"
    else
        message='               '
    endif

    sendcount=len(message)

    call MPI_Bcast(message,sendcount,MPI_CHARACTER,0,MPI_COMM_WORLD,ierr)

    print*, rank, message

    call MPI_Finalize(ierr)

end program 

File: content/courses/parallel-computing-introduction/codes/bcast_ch.cxx
#include <iostream>
#include <iomanip>
#include <cstring>
#include "mpi.h"

using namespace std;

int main(int argc, char *argv[]) {

    int nprocs,rank;

    MPI_Init(&argc,&argv);
    MPI_Comm_size(MPI_COMM_WORLD,&nprocs);
    MPI_Comm_rank(MPI_COMM_WORLD,&rank);

    //extra spece for null terminator
    char message[16];
    strcpy(message,"               ");
    if (rank==0) {
        strcpy(message,"I have a secret");
    }

    MPI_Bcast(message,strlen(message),MPI_CHAR,0,MPI_COMM_WORLD);

    cout<<rank<<" "<<message<<endl;

    MPI_Finalize();

}
        

